added category resolution on compare

scompo opened the pull request:
Created pull request.

SQiShER commented on src/main/java/de/danielbechler/diff/node/DiffNode.java:
```diff
@@ -67,6 +67,7 @@
 	private Class<?> valueType;
 	private TypeInfo valueTypeInfo;
 	private IdentityStrategy childIdentityStrategy;
+	private Set<String> categoriesFromConfig;
```
The name is a bit too specific for my taste. I would prefer `additionalCategories` because it makes the `Set` more versatile.

SQiShER replied:
Oh, and maybe make it `final` and just initialize it with an empty `Set`. That way we never need to worry about NPEs, which are currently possible because the assiciated setter allows to assign `null`.

SQiShER commented on src/main/java/de/danielbechler/diff/node/DiffNode.java:
```diff
@@ -732,6 +740,16 @@ else if (childCount() > 1)
 		return sb.toString();
 	}
 
+	private Set<String> getCategoriesFromConfig() {
```
This getter is never used, so there is no need for it to exist.

SQiShER commented on src/main/java/de/danielbechler/diff/node/DiffNode.java:
```diff
@@ -732,6 +740,16 @@ else if (childCount() > 1)
+	public void setCategoriesFromConfig(Set<String> categoriesFromConfig) {
+
+		this.categoriesFromConfig = categoriesFromConfig;
+	}
```
Unfortunately there is no way to avoid making this method part of the public API, so people will eventually start (ab)using it. That's one of the reasons why I'd prefer a different name for the field, as mentioned above. Although I wouldn't even expose the field via getter or setter at all.

If we can't avoid leaking this detail to the outside, let's make it a cool feature: let's say we call the method `addCategories` and instead of replacing the categories, we just append them to a set. Then people are able to add their own categories later on, when traversing the result. That may allow for some advanced filtering based on criteria that go far beyond the things the ObjectDiffer is (or should be) capable of.

Calling it `addCategories` suggests the direct relationship with `getCategories` and by making it append-only people won't be confused when they try to remove a category that comes from the annotation or the parent node.

I imagine something like this:

```java
public final void addCategories(final Collection<String> additionalCategories)
{
	Assert.notNull(additionalCategories, "additionalCategories");
	this.additionalCategories.addAll(additionalCategories);
}
```

In this example I changed the parameter type from `Set` to `Collection`, because the underlying object is a `Set` anyway so we don't care what collection type they pass.

SQiShER commented:
Thanks for the PR! Looks pretty good so far. I added some thoughts via line comments above.

scompo commented:
I've implemented the changes you suggested in the new commit, also added a couple of tests to verify the methods are working as expected.

scompo commented:
The test `categories should not be modifiable by a client directly` it's not working properly. Calling `visit` on the DiffNode with 2 different `Visitor`s resets the categories to the correct values.
Do you think we need to create an unmodifiable copy of the returned `Set` to prevent changes by editing the returned set or this could be expected?

SQiShER commented on src/main/java/de/danielbechler/diff/ObjectDifferBuilder.java:
```diff
@@ -160,7 +160,8 @@ public ObjectDiffer build()
 				returnableNodeService,
-				introspectionService);
+				introspectionService,
+				inclusionService);
```
You can just inject the `categoryService` here. That way the InclusionService doesn't need to implement the `CategoryResolver` interface.

SQiShER commented on src/main/java/de/danielbechler/diff/inclusion/InclusionService.java:
```diff
@@ -24,13 +24,14 @@
 
 @SuppressWarnings("OverlyComplexAnonymousInnerClass")
-public class InclusionService implements InclusionConfigurer, IsIgnoredResolver
+public class InclusionService implements InclusionConfigurer, IsIgnoredResolver, CategoryResolver
```
There is no need to add the `CategoryResolver` interface here, because that's the purpose of the `CategoryService`.

SQiShER commented:
I like the solution with the unmodifiable set. That way nobody will start to rely on being able to modify the set directly and run into trouble later on. :+1:

scompo commented:
Implemented the injection of `CategoryService` directly in `ObjectDifferBuilder` and removed the `resolveCategories` method from `InclusionService`. I wasn't really sure making `InclusionService` implement `CategoriesReslover` was the right thing to do, in fact this makes much more sense!

SQiShER commented:
Cool, looks pretty good now. I'll take a closer look at the tests now and add my thoughts as I go.

SQiShER commented on src/integration-test/java/de/danielbechler/diff/CategoriesTestIT.groovy:
```diff
@@ -0,0 +1,133 @@
+package de.danielbechler.diff
```
Let's move this to a new package `categories`, since all the other major configuration sections got their own one as well.

SQiShER commented on src/integration-test/java/de/danielbechler/diff/CategoriesTestIT.groovy:
```diff
@@ -0,0 +1,133 @@
+			categoriesMapVisitor.mapCategories.get("thirdString") == ["cat1", "catAnnotation", "addedWhileVisiting"] as Set
+	}
+
+	def "categories should not be modifiable by a client directly"(){
```
I think this one would be better off as a simple unit test in `de.danielbechler.diff.node.DiffNodeTest`.

SQiShER commented on src/integration-test/java/de/danielbechler/diff/CategoriesTestIT.groovy:
```diff
@@ -0,0 +1,133 @@
+			thrown UnsupportedOperationException
+	}
+
+	def "should throw exception when added a null collection"(){
```
Same here: better off as a unit test.

SQiShER commented on src/integration-test/java/de/danielbechler/diff/CategoriesTestIT.groovy:
```diff
@@ -0,0 +1,133 @@
+			categoriesMapVisitor.mapCategories.get("firstString") == ["cat1"] as Set
+			categoriesMapVisitor.mapCategories.get("secondString") == ["cat1"] as Set
+			categoriesMapVisitor.mapCategories.get("thirdString") == ["cat1", "catAnnotation"] as Set
+	}
```
How about doing this instead of using the mapVisitor?

```groovy
node.getChild("firstString").getCategories() == ["cat1"] as Set
...
```

Or was there a specific reason for the visitor approach?

SQiShER commented on src/integration-test/java/de/danielbechler/diff/CategoriesTestIT.groovy:
```diff
@@ -0,0 +1,133 @@
+			this.thirdString = thirdString
+		}
+
+		def getFirstString() {
```
By the way: these getters and setters are automatically generated by Groovy. The only ones you actually need are those for the thirdString due to the annotation.

scompo commented:
Didn't know about groovy generating the getters and setters, that's cool! I just used a java class and adapted it.
Moved the other tests inside `DiffNodeTest`, the logic is in `DiffNode` so I think that the correct place for the tests.
The visitor was used because I used the code from a junit test I made to verify a visitor worked, actually makes more sense using directly `getChildren` and not overcomplicating the test.

SQiShER commented on src/integration-test/java/de/danielbechler/diff/categories/CategoriesTestIT.groovy:
```diff
@@ -0,0 +1,49 @@
+			def node = differ.compare(obj1,obj2)
+		expect :
+			node.getChild("firstString").getCategories() == ["cat1"] as Set
+			node.getChild("secondString").getCategories() == ["cat1"] as Set
```
If there is no distinction between `firstString` and `secondString` I'd rather get rid one of them entirely to strip the test to its bare essentials.

SQiShER commented on src/integration-test/java/de/danielbechler/diff/categories/CategoriesTestIT.groovy:
```diff
@@ -0,0 +1,49 @@
+
+	def obj1 = new MyObject("aaa","aaa", "aaa")
+	def obj2 =  new MyObject("bbb","bbb", "bbb")
+	def differ = ObjectDifferBuilder.startBuilding()
```
Since there is only one test left, the setup can be moved into the test method, so it becomes self-contained.

SQiShER commented on src/test/java/de/danielbechler/diff/node/DiffNodeTest.groovy:
```diff
@@ -245,4 +245,32 @@ class DiffNodeTest extends Specification {
 		  doesOrDoesNotImplement = expectedResult ? 'implements' : 'does not implement'
 	}
+
+	def "should return categories added when visiting"(){
```
The test name doesn't really match what's tested anymore.

SQiShER commented:
> Didn't know about groovy generating the getters and setters, that's cool!

Yeah, Groovy has a lot of magic like that. That's why I like it especially for testing. Helps to be very concise and write very expressive tests. :smiley:

SQiShER commented:
Alright, the few remarks I made over the last minutes aside, I think we're ready to merge. :+1: 

Do all the polishing you like and once you're done, please [squash your commits into a single one](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#squash-your-changes).

SQiShER commented on src/main/java/de/danielbechler/diff/node/DiffNode.java:
```diff
@@ -32,12 +32,7 @@
-import java.util.Map;
-import java.util.Set;
-import java.util.TreeSet;
+import java.util.*;
```
I'd prefer explicit imports here.

SQiShER commented on src/main/java/de/danielbechler/diff/inclusion/InclusionService.java:
```diff
@@ -24,6 +24,7 @@
 
 import java.util.Collection;
 import java.util.LinkedList;
+import java.util.Set;
```
Since this file doesn't contain any actual changes anymore, could you please bring it back into its original condition, so it doesn't show up in the diff?

scompo commented:
Thanks for the remarks, I've pushed the changes you suggested.
I'm learning how to squash commits now :)

scompo commented:
I've created this fresh PR to merge the changes: #160

SQiShER commented:
Closing this one in favor of the follow-up PR #160.
